src/Examine/LuceneEngine/SearchCriteria/LuceneSearchCriteria.cs

```diff
@@ -78,6 +78,10 @@ internal LuceneSearchCriteria(string type, Analyzer analyzer, string[] fields, b
             this.BooleanOperation = occurance;
             this.QueryParser = new MultiFieldQueryParser(_luceneVersion, fields, analyzer);
             this.QueryParser.SetAllowLeadingWildcard(allowLeadingWildcards);
+
+            //see https://lists.gt.net/lucene/java-user/92194
+            QueryParser.SetMultiTermRewriteMethod(MultiTermQuery.SCORING_BOOLEAN_QUERY_REWRITE);
```

Comment:
What impact does this have on other queries? The convo on the mail list is not very helpful https://lists.gt.net/lucene/java-user/92194 

Do you know what he means when this is mentioned?

> Thanks. So the normal way of doing this is setting the rewrite method to 
scoring, and if BooleanQuery.TooManyClauses is catched then switch back to 
constant score method?

Reply:
According to the docs: https://lucene.apache.org/core/4_3_0/core/org/apache/lucene/search/MultiTermQuery.html

> NOTE: if setRewriteMethod(org.apache.lucene.search.MultiTermQuery.RewriteMethod) is either CONSTANT_SCORE_BOOLEAN_QUERY_REWRITE or SCORING_BOOLEAN_QUERY_REWRITE, you may encounter a BooleanQuery.TooManyClauses exception during searching, which happens when the number of terms to be searched exceeds BooleanQuery.getMaxClauseCount(). Setting setRewriteMethod(org.apache.lucene.search.MultiTermQuery.RewriteMethod) to CONSTANT_SCORE_FILTER_REWRITE prevents this.

> The recommended rewrite method is CONSTANT_SCORE_AUTO_REWRITE_DEFAULT: it doesn't spend CPU computing unhelpful scores, and it tries to pick the most performant rewrite method given the query. If you need scoring (like FuzzyQuery, use MultiTermQuery.TopTermsScoringBooleanQueryRewrite which uses a priority queue to only collect competitive terms and not hit this limitation. Note that org.apache.lucene.queryparser.classic.QueryParser produces MultiTermQueries using CONSTANT_SCORE_AUTO_REWRITE_DEFAULT by default.

Reply:
It does seem like 'normal' practice in many examples on the net to use `SetMultiTermRewriteMethod(MultiTermQuery.SCORING_BOOLEAN_QUERY_REWRITE);` and nobody seems to be checking for this exception so not sure. I guess by default we are using MultiFieldQueryParser so it seems like this should just be set.  I wonder if there's a way to pass in a boolean to the searcher to enable/disable this

Reply:
And the comments about this in the .NET source is:

```
    /// <summary> By default QueryParser uses <see cref="F:Lucene.Net.Search.MultiTermQuery.CONSTANT_SCORE_AUTO_REWRITE_DEFAULT" />
    /// when creating a PrefixQuery, WildcardQuery or RangeQuery. This implementation is generally preferable because it
    /// a) Runs faster b) Does not have the scarcity of terms unduly influence score
    /// c) avoids any "TooManyBooleanClauses" exception.
    /// However, if your application really needs to use the
    /// old-fashioned BooleanQuery expansion rewriting and the above
    /// points are not relevant then use this to change
    /// the rewrite method.
    /// </summary>
```

Reply:
I suspect if using something like similarity plugin which is lucene.net contrib you could end up with this error.  It may be worth asking Itamar on stackoverflow although he may just say update to latest lucene.net. Alternatively if you make it a config setting so you can turn it on / off so if anyone does have issue they can just turn it off.

Reply:
I might post on the lucene.net dev mailing list to see what comes back but I had a thought about this. In reality, there is no reason to set this switch if you are not ordering by score. I think that might be why it's disabled by default because there's not implicit requirement that you need score. Might be able to also do that check, if sorting isn't by Score than we don't set the switch

Reply:
That's a good point, I'm guessing you can check at that point if user has set a sort by field?

Reply:
I've changed where this logic is implemented, see rev: e4fb2343d003f31ee66632d96ca44807ff139ab6

Based on that, you could have in fact done this without modifying the core  and used your own implementation of BaseLuceneSearcher by inheriting from the default LuceneSearcher and then override `ISearchResults Search(ISearchCriteria searchParams, int maxResults)` to make this change.

All good, though I've made the change in core and this switch is only applied if ordering by Score.  Will try to add some tests.

Reply:
Hrm, maybe not. I've added tests to support this but it seems setting this switch after you've used the QueryParser is too late, it needs to be set before any queries are parsed unfortunately. I'll see what can be done.

Reply:
I came up with a different way to achieve this, see rev: e4fb2343d003f31ee66632d96ca44807ff139ab6

the problem with `QueryParser.SetMultiTermRewriteMethod(MultiTermQuery.SCORING_BOOLEAN_QUERY_REWRITE);` is that it needs to be set eagerly before any queries are parsed but we don't have that luxury. Looking through the lucene code there are only 3 types of queries that support this type of rewrite but one is obsolete. So I've created a method to rewrite them lazily before the search takes place - it's when the search takes place that the `Rewrite` methods are called.

This works with the latest unit test I've pushed.

Reply:
I blogged it here https://shazwazza.com/post/how-to-set-rewrite-method-on-queries-lazily-in-lucene/ :)